Feature/#3 add default config file locations

ndoering opened the pull request:
This pull request implements the default paths and filenames for the configuration files of TEP.

This closes issue #3

Commit:
Add first default paths for configuration files

This commit add a server, that should provide a access to the
configuration for TEP. Currently the configuration_server only searches
for the configuration file. It therefore searches the default
directories and gives access to the first found file.

Further development should include the old-style config file directory
on linux systems, normally located in ~/.tep/config.ini .

Commit:
Merge branches 'develop' and 'feature/#3-add-default-config-file-locations' of github.com:j-frost/tep into feature/#3-add-default-config-file-locations

Commit:
Reformatting the code with erlfmt

Commit:
Merge branches 'develop' and 'feature/#3-add-default-config-file-locations' of github.com:j-frost/tep into feature/#3-add-default-config-file-locations

Commit:
Move tests to eunit and exit init() without config

The init/1 function of configuration_server now returns {stop, Reason},
if no valid configuration file was given or no valid configuration files
are found in the default locations.

This commit also moves the testing for configuration_server from common
tests to eunit test cases, since it only tests the callback functions
directly and does not need a running gen_server. Since no configuration
file can explicitly chosen at startup, start_link/0 would fail. All API
functions are thin wrappers around gen_server calls and theredore do not
introduce additional logic.

Commit:
Add tep.ini to default config file names

This commit closes issue #3.

ghost commented:
Interestingly, the formatter seems to delete the last empty line of each file. We might hae to look at that again.

j-frost commented on include/configuration_server.hrl:
```diff
@@ -0,0 +1,18 @@
+%%%-------------------------------------------------------------------
+%%% @doc
+%%%
+%%% @end
```
Fill out or delete?

ndoering replied:
Description is added.

j-frost commented on include/configuration_server.hrl:
```diff
@@ -0,0 +1,18 @@
+%%% @doc
+%%%
+%%% @end
+%%% Created : 15. May 2020 9:54 PM
```
I know IDEs (e.g. IntelliJ) really like to put this kind of stuff in here. I mostly remove it and for other languages I've reconfigured my IDE too. However I mostly work on private, unlicensed code bases. Do we want / need this + author info in individual files for licensing reasons?

ndoering replied:
I have removed the creation date. I do not think, that this date is necessary. For the information on the author, I do not have an opinion at the moment.

j-frost commented on include/configuration_server.hrl:
```diff
@@ -0,0 +1,18 @@
+%%% @end
+%%% Created : 15. May 2020 9:54 PM
+%%%-------------------------------------------------------------------
+%% @doc This macro defines the possible file names for the configuration files.
```
Wording: _Accepted_ file names?

ndoering replied:
Good choice. I will fix this.

j-frost commented on include/configuration_server.hrl:
```diff
@@ -0,0 +1,18 @@
+    "tep.ini"
+]).
+
+%% @doc This macro defines the directories in which the config file is searched.
```
Passive voice? "This macro defines where tep will search for its config files".

ndoering replied:
I have taken your suggestion. This will be fixed.

j-frost commented on include/configuration_server.hrl:
```diff
@@ -0,0 +1,18 @@
+%% @doc This macro defines the directories in which the config file is searched.
+-define(CONFIG_PATHS, [
+    filename:absname(""),
+    filename:basedir(user_config, "tep/"),
```
This is so helpful, thanks for finding this :)

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,155 @@
+%%%-------------------------------------------------------------------
+%%% @doc
+%%%
+%%% @end
+%%% Created : 15. May 2020 6:49 PM
+%%%-------------------------------------------------------------------
```
See above

ndoering replied:
This is also filled out, by now.

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,155 @@
+%%% Export these internal functions if eunit is testing this module
+-ifdef(EUNIT).
+
+-export([default_directories/0, config_file_name/0]).
+
+-endif.
+
+%% API
+-export([
+    start_link/0,
+    get_default_directories/0,
+    get_config_file/0
+]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3]).
```
Inconsistent export list chomping.

ndoering replied:
Agreed. I will fix this.

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,155 @@
+-spec get_default_directories() ->
+    {ok, [file:filename_all()]}.
+get_default_directories() ->
+    gen_server:call(configuration_server, default_directories).
+
+%% @doc Returns the used configuration file
+-spec get_config_file() ->
+    {ok, file:filename_all()} | {error, Reason :: term()}.
+get_config_file() ->
+    gen_server:call(configuration_server, config_file).
```
Not sure I agree with the names of the calls here. Is the convention to call for a noun/object that I want to have returned? I'm used to OOP principles where things I want done are verbs.

ndoering replied:
What would you suggest?

ndoering replied:
I have changed this to `choose_config_file/0`. Whats your opinion on that?

j-frost replied:
We're still calling the `configuration_server` here for a `config_file`. I'd have expected something like `get_config_file_path` as the message to `configuration_server` (although maybe that's too verbose).

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,155 @@
+    {ok, State :: #configuration{}, timeout() | hibernate} |
+    {stop, Reason :: term()} | ignore.
+init([]) ->
+    case config_file_name() of
+        {ok, ConfigFilePath} -> {ok, #configuration{config_file_path = ConfigFilePath}};
+        {error, Reason} -> {stop, Reason}
+    end.
```
Once again I have no convention to fall back on for this. In many languages I know people like to reduce both `if`s and `switch`-`case`s. Does Erlang favour them instead? Or should this be a guard clause (`when`)? It feels somewhat clumsy as is, but maybe that's just me lacking experience in Erlang.

ndoering replied:
As per code-analyzer elvis, it seems, that `case`s are prefered over `if`s. I currently do not see where a guard clause could fit. Do you have a suggestion?

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,155 @@
+%% directories and uses the first found file.
+-spec config_file_name() ->
+    {ok, FileName :: file:filename_all()} | {error, Reason :: term()}.
+config_file_name() ->
+    CandidateNames =
+        [
+            filename:absname_join(Path, File)
+            || Path <- default_directories(), File <- ?FILE_NAMES
+        ],
+    case [File || File <- CandidateNames, filelib:is_regular(File)] of
+        ConfigFiles when length(ConfigFiles) >= 1 -> {ok, lists:nth(1, ConfigFiles)};
+        _ -> {error, {no_file, "No config file found"}}
+    end.
```
```Erlang
    [FirstMatch, _] =[
            filename:absname_join(Path, File) ||
            Path <- default_directories(), File <- ?FILE_NAMES, filelib:is_regular(File)
    ]
```
Maybe? Same question here about `case`s.

ndoering replied:
```erlang
[FirstFile | _] = [
            filename:absname_join(Path, File) ||
            Path <- default_directories(), File <- ?FILE_NAMES, filelib:is_regular(File)
]
```
This would match for a list, with at least one element, but fail with an exception, if the input is an empty list. To counter this, I could use a try-catch-block and return the `{error, ...}` in case of an empty list. But that would be very ugly, in my opinion.

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,155 @@
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+%% @doc This function returns the predefined configuration directories.
+-spec default_directories() ->
+    [file:filename_all()].
+default_directories() ->
+    ?CONFIG_PATHS.
```
Is this a best practice? Does `spec` not exist for macros? I don't fully understand why we need this if we're just going to return `id/1`.

ndoering replied:
To the best of my knowledge, there are nor specs for macros, since macros are C-like test replacements (nothing fancy, like Lisp macros). The implementation might change, but this function should return a list of filenames/directories.

j-frost replied:
Also kinda bugs me that there's another pair of functions named with a noun.

j-frost commented on test/configuration_server_test.erl:
```diff
@@ -0,0 +1,24 @@
+configuration_server_test_() ->
+    [
+        ?_assertEqual(?CONFIG_PATHS, configuration_server:default_directories()),
+        ?_assertEqual(
+            {error, {no_file, "No config file found"}},
+            configuration_server:config_file_name()
+        ),
+        ?_assertEqual(
+            {stop, {no_file, "No config file found"}},
+            configuration_server:init([])
+        )
```
Considering this is a unit test I'm not too sure about these. Especially given that they're dependent on the system they're running on. Probably move them to a `SUITE`; either way I think we ought to mock the `?CONFIG_PATHS` if that's possible in Erlang.

ndoering replied:
Maybe we could try to mock the functions in  `?CONFIG_PATH` and the necessary functions from `filelib` and `filenames` with the mocking library **[meck](http://eproxus.github.com/meck)**. Whats your opinion on that?

ndoering replied:
I have added a simple mocking version for the `config_file_name()` function.

j-frost reviewed (changes requested):
Tests break it for me. Seem to invite flip flopping. 

Apart from that, I have a lot of questions because I am noob :)

ghost commented:
> Tests break it for me. Seem to invite flip flopping.

What ist it exactly, that breaks?

Commit:
Add meck mocking library to rebar.config

Commit:
Add descriptions for the configuration module

Commit:
Simple mocking for the config_file_name() function

Commit:
Add mocking for configuration_server:init/1

To test the behaviour of the configuration_server:init/1 function, this
mocking implementation tests the values that should be returned if no
configuration file could be found and if multiple are found.

Commit:
Remove further file creation date

Commit:
Remove inconsistency in access to defaults

This commit removes the inconsistency in the access to the default
names and directories for config files.

Commit:
Rename config_file_name/0 to choose_config_file/0

Commit:
Refactor choose_config_file/2

This commit refactors choose_config_file/0 to a more functional and
therefore more testable choose_config_file/2. The new function takes the
accepted directories and filenames as arguments.

j-frost commented on include/configuration_server.hrl:
```diff
@@ -0,0 +1,18 @@
+%%%-------------------------------------------------------------------
+%%% @doc
+%%% This header file defines a few macros, that TEP needs to find and
+%%% access a configuration file.
```
We know this is a header file from the file extension and contents. I'd like for us to avoid using "this x", "this y" repeatedly. Golang (as usual) has an opinion on this, which I actually find helpful in this case as a mental model. Their linter will mark comments as warn when they don't start with the name of the thing they're describing. 
Also: no comma before "that" in English.

j-frost replied:
Also also: Shouldn't that fit on a single line?

ndoering replied:
This will be changed shortly

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,160 @@
+%%%-------------------------------------------------------------------
+%%% @doc
+%%% This module implements the server, that reads and stores the
+%%% configuration for TEP. It presents an interface for other
+%%% processes to access the configuration variables.
```
```suggestion
%%% This module implements the server that reads and stores the
%%% configuration for TEP. It presents an interface for other
%%% processes to access configuration variables.
```
Again: Lines seem pretty narrow.

ndoering replied:
This text will be reworked.

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,160 @@
+init([]) ->
+    Directories = default_directories(),
+    FileNames = default_filenames(),
+    case choose_config_file(Directories, FileNames) of
+        {ok, ConfigFilePath} -> {ok, #configuration{config_file_path = ConfigFilePath}};
+        {error, Reason} -> {stop, Reason}
+    end.
```
Still seems to prescriptive/imperative. Compare `bad` on line 17 here: https://github.com/inaka/erlang_guidelines/blob/master/src/smaller_functions.erl

ndoering replied:
This function is completely reworked and uses much more pattern matching.

j-frost commented on include/configuration_server.hrl:
```diff
@@ -0,0 +1,18 @@
+%%% access a configuration file.
+%%% @end
+%%%-------------------------------------------------------------------
+%% @doc This macro defines the accepted file names for the configuration files.
```
```suggestion

%% @doc This macro defines the accepted file names for the configuration files.
```
for readability

ndoering replied:
This will be automatically removed by the formatter `rebar3 fmt`.

j-frost replied:
I see. In that case I guess we'll leave it for now and try to grow accustomed, since it's so opinionated. I could see this becoming annoying however...

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,160 @@
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+%% @doc Spawns the server and registers the local name (unique)
```
```suggestion
%%%===================================================================

%% @doc Spawns the server and registers the local name (unique)
```

ndoering replied:
This will be automatically removed by the formatter `rebar3 fmt`.

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,160 @@
+get_config_file() ->
+    gen_server:call(configuration_server, config_file).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+%% @private
+%% @doc Initializes the server
```
```suggestion
%%%===================================================================
%%% gen_server callbacks
%%%===================================================================

%% @private
%% @doc Initializes the server
```

ndoering replied:
This will be automatically removed by the formatter `rebar3 fmt`.

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,160 @@
+    FileNames :: [file:filename_all()]
+) ->
+    {ok, FileName :: file:filename_all()} | {error, Reason :: term()}.
+choose_config_file(Directories, FileNames) ->
```
As the doc says the function's searching, I'd suggest renaming this to `find_config_file` or similar.

ndoering replied:
This is reworked. Naming has changed

j-frost commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,160 @@
+choose_config_file(Directories, FileNames) ->
+    CandidateNames =
+        [filename:absname_join(Path, File) || Path <- Directories, File <- FileNames],
+    case [File || File <- CandidateNames, filelib:is_regular(File)] of
+        ConfigFiles when length(ConfigFiles) >= 1 -> {ok, lists:nth(1, ConfigFiles)};
+        _ -> {error, {no_file, "No config file found"}}
+    end.
```
Again I'm reminded of https://github.com/inaka/erlang_guidelines/blob/master/src/smaller_functions.erl

j-frost replied:
Either way I personally don't find this very readable. So even if the `case` is something we can't get rid of, I think this is too terse as is.

ndoering replied:
This has been changed to work without `case`.

j-frost commented on test/configuration_server_test.erl:
```diff
@@ -0,0 +1,40 @@
+
+-include("configuration_server.hrl").
+
+default_directories_test() ->
+    ?assertEqual(?CONFIG_PATHS, configuration_server:default_directories()).
+
+default_filenames_test() ->
+    ?assertEqual(?FILE_NAMES, configuration_server:default_filenames()).
```
I'm the biggest proponent of TDD, but this just strikes me as awkward. This is basically testing implementation details, which is an anti-pattern imho.

ndoering replied:
Ok, i'll drop it.

j-frost commented on test/configuration_server_test.erl:
```diff
@@ -0,0 +1,40 @@
+default_filenames_test() ->
+    ?assertEqual(?FILE_NAMES, configuration_server:default_filenames()).
+
+choose_file_name_test() ->
+    Paths = ["/etc/tep/", "./"],
+    FileNames = ["config.ini", "tep.ini"],
+    meck:new(filelib, [unstick]),
+    meck:expect(filelib, is_regular, fun
+        (Path) when Path == "./tep.ini" -> true;
+        (_) -> false
+    end),
+    ?assertEqual(
+        {ok, "./tep.ini"},
+        configuration_server:choose_config_file(Paths, FileNames)
+    ),
+    ?assert(meck:validate(filelib)),
+    meck:unload().
```
These tests look like they were written after the implementation, if I understand them correctly ;)
We need cases for all behaviour we want prod to exhibit. This just tests for the `./tep.ini` case, right? Since "all cases" here amounts to six instances, I think it's still reasonable to write six individual samples, but if you want we can look into looping tests (which should be easy with eunit). I don't think code generators are warranted just yet.

ndoering replied:
This has been reworked.

j-frost reviewed (changes requested):
Still some stuff I'm either unclear on or would like to see changed. Beginnings are difficult and all that :)

j-frost reviewed (commented):
> > Tests break it for me. Seem to invite flip flopping.
> 
> What ist it exactly, that breaks?

Ah, sorry. That was my final comment on the initial review. Nothing actually broke, but my review didn't end in me approving it, so I felt like explaining why I was requesting further changes.

Commit:
Refactoring for configuration server

To be more in line with the style preferred in Erlang applications, this
code now is rid of case statements and makes heavy use of pattern
matching.

Commit:
Rewording of the documentation

j-frost commented on src/configuration_server.erl:
```diff
@@ -1,8 +1,8 @@
-%%% processes to access the configuration variables.
+%%% The configuration_server reads and stores the configuration  for
+%%% TEP. It presents an interface for other processes to access
+%%% configuration variables.
```
So `erlfmt` will remove empty lines between two different types of comments, but it won't touch this. I just put everything on one line and ran `rebar3 fmt`, changed nothing. That's kind of annoying, and we'll have to keep that in mind moving forward. I guess `erlfmt`'s opinion in this case is "this space is for prose", so it tries not to get in the way.

j-frost commented on src/configuration_server.erl:
```diff
@@ -131,30 +148,64 @@ code_change(_OldVsn, State = #configuration{}, _Extra) ->
+-spec prepare_default_directories() ->
     [file:filename_all()].
-default_directories() ->
+prepare_default_directories() ->
```
So I understand why this exists the way it does, but I cannot in good conscience leave this uncommented. Considering they're not really doing any prep work, I'd rather we stick with the much less exciting `get_default_directories` - but that's taken by the API function already (in both cases, of course). Seems to indicate a code smell to me. I'm not sure, but I think Erlang wants us to treat constants like constants and not wrap them in functions.

j-frost replied:
Maybe we can give them slightly longer / more descriptive names to compensate?

j-frost commented on src/configuration_server.erl:
```diff
@@ -41,13 +52,19 @@ start_link() ->
+-spec get_default_filenames() ->
+    {ok, [file:filename_all()]}.
+get_default_filenames() ->
+    gen_server:call(configuration_server, get_default_filenames).
```
Why do we expose these as API? Are we anticipating using them for debugging? As long as I know what the config says, do I (as an external process) really need to know where it's saved and where it's searched? Shouldn't that be encapsulated in `configuration_server`?

Commit:
Clean up some naming and formatting

see #3

j-frost reviewed (commented):
I just pushed some changes as suggestions of how I think we can clean this up and get it merged.

Commit:
Fix minor spelling mistake

ndoering commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,207 @@
+    {stop, Reason :: term()} | ignore.
+init([]) ->
+    ConfigFilePathSearchResult = search_for_config_file_path(
+        ?TEP_DEFAULT_CONFIG_LOCATIONS,
+        ?TEP_DEFAULT_CONFIG_FILE_NAMES
```
I have a problem with the multiple use of these macros. These macros currently expand only into a list, containing a few function calls. This might change. The impact on the execution would be easier to be accounted for, if it is encapsuled in a  dedicated function. This would also reduce the overhead of multiple copies in the source code and therefore also in the compiled modules. This would in my opinion violate the DRY principle.

ndoering replied:
We checked if this has any effect on the compiled code size, but that is negligable. So for now, lets keep using this.

ndoering commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,207 @@
+        [file:filename_all()],
+    FileNames :: [file:filename_all()]
+) ->
+    {ok,
+        FileName ::
+            file:filename_all()} |
```
This a some weird looking code. That might be better formatted.

ndoering replied:
Just tested with the formatter. That's its doing.

ndoering commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,207 @@
+        ?TEP_DEFAULT_CONFIG_LOCATIONS,
+        ?TEP_DEFAULT_CONFIG_FILE_NAMES
+    ),
+    StoreResult = create_configuration(ConfigFilePathSearchResult),
```
Being a little picky with the names, I would say, that we already know, that this `StoreResult` is a result.In my opinion, there is no reason to name it a `*Result`, same as above. I would name it `Configuration` or `Config`.

ndoering replied:
This will be changed to Configuration

ndoering commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,207 @@
+    {stop, Reason :: term(), NewState :: #configuration{}}.
+handle_call(get_config_file_path, _From, State = #configuration{}) ->
+    {reply, State#configuration.config_file_path, State};
+handle_call(get_default_directories, _From, State = #configuration{}) ->
+    {reply, {ok, ?TEP_DEFAULT_CONFIG_LOCATIONS}, State};
+handle_call(get_default_filenames, _From, State = #configuration{}) ->
+    {reply, {ok, ?TEP_DEFAULT_CONFIG_FILE_NAMES}, State};
```
For the macros, please see above.

ndoering replied:
Checked as seen above.

ndoering commented on src/configuration_server.erl:
```diff
@@ -0,0 +1,207 @@
+    {ok, State :: #configuration{}, timeout() | hibernate} |
+    {stop, Reason :: term()} | ignore.
+init([]) ->
+    ConfigFilePathSearchResult = search_for_config_file_path(
```
This will be changed to `ConfigFilePath`

Commit:
Shorten variable names for better readability
